### PR TITLE
feat: adds support to sample success logs

### PR DESCRIPTION
**Pull request by epinault**
In a real world production system, logs can get really large and having all the success may not be possible or needed. Having a way to sample it can help control the volume of logs

### CODE REVIEW

File: lib/uinta/plug.ex [hunk 1]
```diff
@@ -275,5 +287,18 @@ if Code.ensure_loaded?(Plug) do
     defp query_type("mutation" <> _), do: "MUTATION"
     defp query_type("{" <> _), do: "QUERY"
     defp query_type(_), do: nil
+
+    defp should_log?(ratio) do
+      random_float() <= ratio
```
[the remark this answers is not in the document]
Comment (epinault) [reply lib/uinta/plug.ex:292]: oops should have been 1.0. let me fix it

File: lib/uinta/plug.ex [hunk 2]
```diff
@@ -109,7 +111,13 @@ if Code.ensure_loaded?(Plug) do
         ignored_paths: Keyword.get(opts, :ignored_paths, []),
         include_unnamed_queries: Keyword.get(opts, :include_unnamed_queries, false),
         include_variables: Keyword.get(opts, :include_variables, false),
-        filter_variables: Keyword.get(opts, :filter_variables, @default_filter)
+        filter_variables: Keyword.get(opts, :filter_variables, @default_filter),
+        success_log_sampling_percent:
```
Comment (brettbeatty) [inline lib/uinta/plug.ex:115]: Would ratio be clearer than percent if it's a float?

**Review by brettbeatty — APPROVED** [review]

File: lib/uinta/plug.ex (hunk 2 above)
Comment (epinault) [reply lib/uinta/plug.ex:115]: yea I was debating . I ll change it to ratio

**Review by brettbeatty — APPROVED** [review]

File: lib/uinta/plug.ex [hunk 3]
```diff
@@ -118,7 +126,11 @@ if Code.ensure_loaded?(Plug) do
       start = System.monotonic_time()
 
       Conn.register_before_send(conn, fn conn ->
-        if conn.request_path not in opts.ignored_paths || conn.status >= 300 do
+        # Log successful request if path is not filtered based on the sampling pool
+        # or log all HTTP status >= 300 (usually errors)
+        if (conn.request_path not in opts.ignored_paths &&
+              should_log?(opts[:success_log_sampling_ratio])) ||
```
Comment (dnsbty) [inline lib/uinta/plug.ex:132]: It feels like with this addition, the complexity of the clause becomes a lot higher. What would you think of moving all of this into a new `should_log_request?/2` that would take the `conn` and the `opts` and return a boolean that encompasses all of these conditions?

File: lib/uinta/plug.ex [hunk 4]
```diff
@@ -275,5 +287,18 @@ if Code.ensure_loaded?(Plug) do
     defp query_type("mutation" <> _), do: "MUTATION"
     defp query_type("{" <> _), do: "QUERY"
     defp query_type(_), do: nil
+
+    defp should_log?(ratio) do
```
Comment (dnsbty) [inline lib/uinta/plug.ex:291]: What would you think of the name `should_include_in_sample?` or something to make it more obvious from the high level what this is checking?

**Review by dnsbty — CHANGES_REQUESTED** [review]

**Pull request merged**